File: ecosystem_client_js/src/discovery.js
const http = require("http");
const https = require("https");

const DiscoveryMode = {
    REGISTRY: "registry",
    PEER_TO_PEER: "peer_to_peer",
    STANDALONE: "standalone",
};

class DiscoveryManager {
    constructor(config) {
        this.config = config;
        this.mode = DiscoveryMode.STANDALONE;
        this._peers = {};
        this._mdnsPeers = [];
    }

    async detectMode() {
        if (!this.config.enabled) {
            this.mode = DiscoveryMode.STANDALONE;
            return this.mode;
        }

        if (await this._checkRegistry()) {
            this.mode = DiscoveryMode.REGISTRY;
            return this.mode;
        }

        const mdnsPeers = this._checkMdns();
        if (mdnsPeers.length > 0) {
            this._mdnsPeers = mdnsPeers;
            this.mode = DiscoveryMode.PEER_TO_PEER;
            return this.mode;
        }

        if (Object.keys(this.config.peers).length > 0) {
            this.mode = DiscoveryMode.PEER_TO_PEER;
            return this.mode;
        }

        this.mode = DiscoveryMode.STANDALONE;
        return this.mode;
    }

    async getPeers() {
        if (this.mode === DiscoveryMode.REGISTRY) {
            return await this._fetchRegistryServices();
        }
        if (this.mode === DiscoveryMode.PEER_TO_PEER) {
            const peers = {};
            for (const [name, url] of Object.entries(this.config.peers)) {
                peers[name] = { name, base_url: url };
            }
            return peers;
        }
        return {};
    }

    async registerSelf(name, host, port, healthEndpoint, webhookUrl, subscriptions) {
        if (this.mode !== DiscoveryMode.REGISTRY) return false;
        try {
            const payload = JSON.stringify({
                name, host, port,
                health_endpoint: healthEndpoint,
                webhook_url: webhookUrl,
                subscriptions: subscriptions || [],
            });
            await this._post(`${this.config.registryUrl}/register`, payload);
            return true;
        } catch {
            return false;
        }
    }

    async deregisterSelf(name) {
        if (this.mode !== DiscoveryMode.REGISTRY) return false;
        try {
            await this._delete(`${this.config.registryUrl}/deregister/${name}`);
            return true;
        } catch {
            return false;
        }
    }

    async _checkRegistry() {
        try {
            await this._get(`${this.config.registryUrl}/health`);
            return true;
        } catch {
            return false;
        }
    }

    _checkMdns() {
        try {
            const { Bonjour } = require("bonjour-service");
            // Synchronous browse not practical; return empty for now
        } catch { /* not installed */ }
        return [];
    }

    async _fetchRegistryServices() {
        try {
            const data = await this._get(`${this.config.registryUrl}/services`);
            const peers = {};
            for (const svc of data) {
                peers[svc.name] = svc;
            }
            return peers;
        } catch {
            return {};
        }
    }

    _get(url) {
        return new Promise((resolve, reject) => {
            const lib = url.startsWith("https") ? https : http;
            const req = lib.get(url, { timeout: this.config.requestTimeout }, (res) => {
                let body = "";
                res.on("data", (chunk) => body += chunk);
                res.on("end", () => {
                    if (res.statusCode >= 400) return reject(new Error(`HTTP ${res.statusCode}`));
                    try { resolve(JSON.parse(body)); } catch { resolve(body); }
                });
            });
            req.on("error", reject);
            req.on("timeout", () => { req.destroy(); reject(new Error("timeout")); });
        });
    }

    _post(url, body) {
        return new Promise((resolve, reject) => {
            const parsed = new URL(url);
            const lib = parsed.protocol === "https:" ? https : http;
            const req = lib.request({
                hostname: parsed.hostname, port: parsed.port, path: parsed.pathname,
                method: "POST", headers: { "Content-Type": "application/json", "Content-Length": Buffer.byteLength(body) },
                timeout: this.config.requestTimeout,
            }, (res) => {
                let data = "";
                res.on("data", (chunk) => data += chunk);
                res.on("end", () => {
                    if (res.statusCode >= 400) return reject(new Error(`HTTP ${res.statusCode}`));
                    try { resolve(JSON.parse(data)); } catch { resolve(data); }
                });
            });
            req.on("error", reject);
            req.on("timeout", () => { req.destroy(); reject(new Error("timeout")); });
            req.write(body);
            req.end();
        });
    }

    _delete(url) {
        return new Promise((resolve, reject) => {
            const parsed = new URL(url);
            const lib = parsed.protocol === "https:" ? https : http;
            const req = lib.request({
                hostname: parsed.hostname, port: parsed.port, path: parsed.pathname,
                method: "DELETE", timeout: this.config.requestTimeout,
            }, (res) => {
                let data = "";
                res.on("data", (chunk) => data += chunk);
                res.on("end", () => resolve(data));
            });
            req.on("error", reject);
            req.on("timeout", () => { req.destroy(); reject(new Error("timeout")); });
            req.end();
        });
    }
}

module.exports = { DiscoveryManager, DiscoveryMode };

File: ecosystem_client/discovery.py
"""Service discovery with three-mode cascade: registry -> mDNS -> static -> standalone."""
from __future__ import annotations

import logging
from enum import Enum
from typing import Any

import httpx

from .config import EcosystemConfig

logger = logging.getLogger(__name__)


class DiscoveryMode(str, Enum):
    REGISTRY = "registry"
    PEER_TO_PEER = "peer_to_peer"
    STANDALONE = "standalone"


class DiscoveryManager:
    """Detects operating mode and resolves peer locations."""

    def __init__(self, config: EcosystemConfig):
        self.config = config
        self._mode: DiscoveryMode | None = None
        self._peers: dict[str, dict[str, Any]] = {}
        self._mdns_peers: list[dict] = []

    @property
    def mode(self) -> DiscoveryMode:
        return self._mode or DiscoveryMode.STANDALONE

    async def detect_mode(self) -> DiscoveryMode:
        """Run the discovery cascade and return the detected mode."""
        if not self.config.enabled:
            self._mode = DiscoveryMode.STANDALONE
            logger.info("Ecosystem disabled — standalone mode")
            return self._mode

        # 1. Check registry
        if await self._check_registry():
            self._mode = DiscoveryMode.REGISTRY
            logger.info(f"Registry found at {self.config.registry_url} — registry mode")
            return self._mode

        # 2. Check mDNS
        self._mdns_peers = self._check_mdns()
        if self._mdns_peers:
            self._mode = DiscoveryMode.PEER_TO_PEER
            logger.info(f"Found {len(self._mdns_peers)} peers via mDNS — peer-to-peer mode")
            return self._mode

        # 3. Check static peers
        if self.config.peers:
            self._mode = DiscoveryMode.PEER_TO_PEER
            logger.info(f"Using {len(self.config.peers)} static peers — peer-to-peer mode")
            return self._mode

        # 4. Standalone
        self._mode = DiscoveryMode.STANDALONE
        logger.info("No registry or peers found — standalone mode")
        return self._mode

    async def get_peers(self) -> dict[str, dict[str, Any]]:
        """Return discovered peers based on current mode."""
        if self._mode == DiscoveryMode.REGISTRY:
            services = await self._fetch_registry_services()
            self._peers = {
                svc["name"]: svc for svc in services
            }
            return self._peers

        if self._mode == DiscoveryMode.PEER_TO_PEER:
            # Merge mDNS and static peers
            peers = {}
            for mdns_peer in self._mdns_peers:
                name = mdns_peer["name"]
                peers[name] = {
                    "name": name,
                    "base_url": f"http://{mdns_peer['host']}:{mdns_peer['port']}",
                }
            for name, url in self.config.peers.items():
                if name not in peers:
                    peers[name] = {"name": name, "base_url": url}
            self._peers = peers
            return self._peers

        # STANDALONE
        return {}

    async def _check_registry(self) -> bool:
        """Check if the ecosystem registry is reachable."""
        try:
            async with httpx.AsyncClient(timeout=self.config.request_timeout) as client:
                resp = await client.get(f"{self.config.registry_url}/health")
                return resp.status_code == 200
        except Exception:
            return False

    def _check_mdns(self) -> list[dict]:
        """Scan for ecosystem services via mDNS. Returns list of peer dicts."""
        try:
            from registry.discovery import EcosystemDiscovery

            disc = EcosystemDiscovery()
            return disc.discover_services()
        except ImportError:
            logger.debug("mDNS discovery not available (zeroconf not installed)")
            return []
        except Exception as e:
            logger.debug(f"mDNS discovery failed: {e}")
            return []

    async def _fetch_registry_services(self) -> list[dict]:
        """Fetch all services from the registry, sorted by priority (highest first)."""
        try:
            async with httpx.AsyncClient(timeout=self.config.request_timeout) as client:
                resp = await client.get(f"{self.config.registry_url}/services")
                resp.raise_for_status()
                services = resp.json()
                # Sort by priority descending so highest-priority peers are first
                services.sort(key=lambda s: s.get("priority", 0), reverse=True)
                return services
        except Exception as e:
            logger.warning(f"Failed to fetch services from registry: {e}")
            return []

    async def register_self(self, name: str, host: str, port: int,
                            health_endpoint: str, webhook_url: str | None = None,
                            subscriptions: list[str] | None = None,
                            priority: int = 0) -> bool:
        """Register this service with the registry (Mode 1 only)."""
        if self._mode != DiscoveryMode.REGISTRY:
            return False
        try:
            payload = {
                "name": name,
                "host": host,
                "port": port,
                "health_endpoint": health_endpoint,
                "webhook_url": webhook_url,
                "subscriptions": subscriptions or [],
                "priority": priority,
            }
            async with httpx.AsyncClient(timeout=self.config.request_timeout) as client:
                resp = await client.post(f"{self.config.registry_url}/register", json=payload)
                resp.raise_for_status()
                logger.info(f"Registered with ecosystem registry as '{name}'")
                return True
        except Exception as e:
            logger.warning(f"Failed to register with registry: {e}")
            return False

    async def deregister_self(self, name: str) -> bool:
        """Deregister this service from the registry (Mode 1 only)."""
        if self._mode != DiscoveryMode.REGISTRY:
            return False
        try:
            async with httpx.AsyncClient(timeout=self.config.request_timeout) as client:
                resp = await client.delete(f"{self.config.registry_url}/deregister/{name}")
                return resp.status_code < 400
        except Exception:
            return False
